iothub_client/samples/pnp/common/pnp_protocol_helpers.c

```diff
@@ -0,0 +1,327 @@
+            {
+                // If the child element is NOT an object OR its not a model the application knows about, this is a property of the model's root component.
+                // Invoke the application's passed in callback for it to process this property.
+                pnpPropertyCallback(NULL, name, value, version, userContextCallback);
+            }
+        }
+
+        result = true;
+    }
+
+    return result;
+}
+
+//
+// GetDesiredJson retrieves JSON_Object* in the JSON tree corresponding to the desired payload.
+//
+static JSON_Object* GetDesiredJson(DEVICE_TWIN_UPDATE_STATE updateState, JSON_Value* rootValue)
+{
+    JSON_Object* rootObject = NULL;
+    JSON_Object* desiredObject;
+
+    if ((rootObject = json_value_get_object(rootValue)) == NULL)
+    {
+        LogError("Unable to get root object of JSON");
+        desiredObject = NULL;
+    }
+    else
+    {
+        if (updateState == DEVICE_TWIN_UPDATE_COMPLETE)
+        {
+            // For a complete update, the JSON from IoTHub will contain both "desired" and "reported" - the full twin.
+            // We only care about "desired" in this sample, so just retrieve it.
+            desiredObject = json_object_get_object(rootObject, g_JSONDesiredName);
+        }
+        else
+        {
+            // For a patch update, IoTHub does not explicitly put a "desired:" JSON envelope.  The "desired-ness" is implicit 
+            // in this case, so here we simply need the root of the JSON itself.
+            desiredObject = rootObject;
+        }
+    }
+
+    return desiredObject;
+}
+
+bool PnPHelper_ProcessTwinData(DEVICE_TWIN_UPDATE_STATE updateState, const unsigned char* payload, size_t size, const char** componentsInModel, size_t numComponentsInModel, PnPHelperPropertyCallbackFunction pnpPropertyCallback, void* userContextCallback)
+{
+    char* jsonStr = NULL;
+    JSON_Value* rootValue = NULL;
+    JSON_Object* desiredObject;
+    bool result;
+
+    if ((jsonStr = PnPHelper_CopyPayloadToString(payload, size)) == NULL)
+    {
+        LogError("Unable to allocate twin buffer");
+        result = false;
+    }
+    else if ((rootValue = json_parse_string(jsonStr)) == NULL)
+    {
+        LogError("Unable to parse device twin JSON");
+        result = false;
+    }
+    else if ((desiredObject = GetDesiredJson(updateState, rootValue)) == NULL)
+    {
+        LogError("Cannot retrieve desired JSON object");
+        result = false;
+    }
+    else
+    {
+        // Visit each sub-element in the desired portion of the twin JSON and invoke pnpPropertyCallback as appropriate.
+        result = VisitDesiredObject(desiredObject, componentsInModel, numComponentsInModel, pnpPropertyCallback, userContextCallback);
+    }
+
+    json_value_free(rootValue);
+    free(jsonStr);
+
+    return result;
+}
+
+char* PnPHelper_CopyPayloadToString(const unsigned char* payload, size_t size)
+{
+    char* jsonStr;
+
+    if ((jsonStr = (char*)malloc(size + 1)) == NULL)
```

Comment:
"size + 1" to a var, so that it's not recomputed again for the error msg

Reply:
Fixing.